paddle/cinn/ir/ir_schedule_error.h

```diff
@@ -0,0 +1,173 @@
+  return (status == 0) ? res.get() : name;
+}
+#else
+inline std::string demangle(std::string name) { return name; }
+#endif
+
+static std::string GetErrorSumaryString(const std::string& what,
+                                        const char* file,
+                                        int line) {
+  std::ostringstream sout;
+  sout << "\n----------------------\nError Message "
+          "Summary:\n----------------------\n";
+  sout << what << "(at " << file << " : " << line << ")" << std::endl;
+  return sout.str();
+}
+
+static std::string GetCurrentTraceBackString() {
+  std::ostringstream sout;
+  sout << "\n\n--------------------------------------\n";
+  sout << "C++ Traceback (most recent call last):";
+  sout << "\n--------------------------------------\n";
+#if !defined(_WIN32) && !defined(PADDLE_WITH_MUSL)
+  static constexpr int TRACE_STACK_LIMIT = 100;
+
+  void* call_stack[TRACE_STACK_LIMIT];
+  auto size = backtrace(call_stack, TRACE_STACK_LIMIT);
+  auto symbols = backtrace_symbols(call_stack, size);
+  Dl_info info;
+  int idx = 0;
+  int end_idx = 0;
+  for (int i = size - 1; i >= end_idx; --i) {
+    if (dladdr(call_stack[i], &info) && info.dli_sname) {
+      auto demangled = demangle(info.dli_sname);
+      std::string path(info.dli_fname);
+      // C++ traceback info are from core.so
+      if (path.substr(path.length() - 3).compare(".so") == 0) {
+        sout << idx++ << " " << demangled << "\n";
+      }
+    }
+  }
+  free(symbols);
+#else
+  sout << "Not support stack backtrace yet.\n";
+#endif
+  return sout.str();
+}
+
+static std::string GetTraceBackString(const std::string& what,
+                                      const char* file,
+                                      int line) {
+  return GetCurrentTraceBackString() + GetErrorSumaryString(what, file, line);
+  // return GetErrorSumaryString(what, file, line);
+}
+
+struct EnforceNotMet : public std::exception {
+ public:
+  EnforceNotMet(const std::string& str, const char* file, int line)
+      : err_str_(GetTraceBackString(str, file, line)) {}
+
+  const char* what() const noexcept override { return err_str_.c_str(); }
+
+ private:
+  std::string err_str_;
+};
+
+#define CINN_THROW(...)                                              \
+  do {                                                               \
+    try {                                                            \
+      throw enforce::EnforceNotMet(__VA_ARGS__, __FILE__, __LINE__); \
+    } catch (const std::exception& e) {                              \
+      std::cout << e.what() << std::endl;                            \
+      throw;                                                         \
+    }                                                                \
+  } while (0)
+}  // namespace enforce
+
+/**
+ *  \brief Indicates the level of printing error message in the current Schedule
+ */
+enum class ScheduleErrorMessageLevel : int32_t {
```

Comment:
Can the ErrorMessage use at other places? If so, should we remove `Schedule` in naming? Then we can use  the code you implemented all over the CINN!

Reply:
Technically no problem. But  its namespace is `cinn::ir` now, i don't know if it is convenient/necessary to use this in the same level of `ir`, like `hlir`, `optim`, etc.

Reply:
What about changing the name and the location of this file when we have the need later?

Reply:
If so, can we change the namespace, location of file in this PR? We would like to reuse for other places.